Add a new post - Understanding Index Request Cache (#3361)
opensearch-project/project-website: _posts/2024-10-01-understanding-index-request-cache.markdown

---
layout: post
title: "Optimizing query performance: The implementation of index request cache in OpenSearch" 
authors:
   - awskiran
   - kkhatua
   - upasagar
   - kolchfa
date: 2024-10-01 
categories:
  - technical-posts
  - search
meta_keywords: OpenSearch cluster, Opensearch caching, index request cache, search performance optimization, search latency
meta_description: 
---

Speed and efficiency are essential for search users. OpenSearch achieves these through various mechanisms, one of the most important of which is [index request cache](https://opensearch.org/docs/latest/search-plugins/caching/request-cache/). This blog post describes how this cache is implemented and how it optimizes query performance.

## What is index request cache?

Index request cache is designed to accelerate search queries in OpenSearch by storing query results at the shard level, as shown in the following image. 

> ![Index-Request-Cache](/assets/media/blog-images/2024-10-01-understanding-index-request-cache/cache_location.png){:class="img-centered" style="width:750px;"}

Reviewer: In this image, please change "Opensearch" to "OpenSearch".

Author: Done, good call !

## Caching policy

Not all search requests are eligible for caching in index request cache. Search requests which specify `size=0` are cached by default. These requests cache only metadata, such as the total number of results or hits. 

The following requests are ineligible for caching:

* **Non-deterministic requests:** Searches involving functions like `Math.random()` or relative times such as `now` or `new Date()`.
* **Scroll and Profile API requests**
* **DFS query then fetch requests:** This search type depends on both index content and overridden statistics, leading to inaccurate scores when the statistics differ (for example, because of shard updates).

You can enable caching for individual search requests by setting the `request_cache` query parameter to `true`:

```json
GET /students/_search?request_cache=true
{
  "query": {
    "match": {
      "name": "doe john"
    }
  }
}
```

## Understanding cache entries

Each cache entry is a key-value pair of **Key → BytesReference**.

> A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:

Author: I think this should be a `Key` instead of `key` because i want the reader to think of this as a java class which follows the standard of camelcase and its easier to relate with the diagram that also is in the same format.

> A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:

Reviewer: @kiranprakash154 Ok, that works for me.

> A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:

Reviewer 2: Yes, should be capital. Classes in Java are capitalized.

> 1. **CacheEntity:** An [IndexShardCacheEntity](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesService.java#L1866C24-L1866C45) contains the [IndexShard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java). This reference links the key to the shard it belongs to.

Author: yep, i love that !

The following is the implementation of the `getOrCompute` method:

```javascript
function getOrCompute(CacheEntity, DirectoryReader, cacheKey) {
    // Step 1: Get the current state identifier of the shard
    readerCacheKeyId = DirectoryReader.getDelegatingCacheKey().getId()

    // Step 2: Create a unique key for the cache entry
    key = new Key(CacheEntity, cacheKey, readerCacheKeyId)

    // Step 3: Check if the result is already in the cache
    value = cache.computeIfAbsent(key)
    
    // Step 4: If the result was computed (not retrieved from the cache), register a cleanup listener
    if (cacheLoader.isLoaded()) {
        cleanupKey = new CleanupKey(CacheEntity, readerCacheKeyId)
        OpenSearchDirectoryReader.addReaderCloseListener(DirectoryReader, cleanupKey)
    }

    // Step 5: Return the cached or computed result
    return value
}
```

## Cache invalidation

An `IndexReader` provides a point-in-time view of an index. Any operation that modifies the index's contents creates a new `IndexReader` and closes the old one. Cache entries created by the old `IndexReader` then become stale and must be cleaned up.

### CleanupKey

> When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`

Reviewer: ```suggestion
When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`.
```

> 

Author: ya good call, its actually referenced in the second image from here. I'll remove `as shown in the following diagram.`

> 

Reviewer: Ideally, we should leave "as shown in the following diagram" but accurately describe what the diagram shows.

> 

Reviewer 2: Can we define CleanupKey here as the first sentence? Something like:
```
A CleanupKey corresponds to a Key to be deleted. The relationship between Key and CleanupKey is shown on the following diagram.
<diagram>
When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`.
```

> 

Reviewer: "in" the following diagram. Should the instances of "Key" and "CleanupKey" in the first sentence be in code font?

> 

Author: I have changed it to 
"A `CleanupKey` corresponds to a `Key` to be deleted. The relationship between them is shown in the following diagram."

The third entity in the Key class, **BytesReference**, is not used in `CleanupKey` because it represents the cached data itself, which is not needed to identify which entries should be cleaned up. `CleanupKey` is only concerned with identifying the entries to remove, not their contents.

A cache entry can become invalid because of the following operations:

- **Refresh/merge**: A `refresh` or `merge` operation creates a new `IndexReader`, thereby invalidating the cache entry.
- **Clear cache**: A clear cache API operation invalidates all request cache entries for the specified index. You can call the Clear Cache API as follows:
  ```bash
  POST /my-index/_cache/clear?request=true
  ```

Any scenario that invalidates an `IndexReader` or explicitly clears the cache adds corresponding `CleanupKeys` to a collection called `KeysToClean`, as shown in the following diagram.

![Cache-Clear](/assets/media/blog-images/2024-10-01-understanding-index-request-cache/keys_to_clean_insert.png){:class="img-centered" style="width:650px;"}

## Cache cleanup

> OpenSearch runs a background job called [CacheCleaner](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesService.java#L1678) every minute on a separate thread. This job calls the [`cleanCache`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L698) method, which iterates through all cache entries, maps each `Key` to a `CleanupKey` in `KeysToClean`, and removes the corresponding entries, as shown in the following diagram.

Reviewer: ```suggestion
Every minute, OpenSearch runs a background job called [CacheCleaner](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesService.java#L1678) on a separate thread. This job calls the [`cleanCache`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L698) method, which iterates through all cache entries, maps each `Key` to a `CleanupKey` in `KeysToClean`, and removes the corresponding entries, as shown in the following diagram.
```

![Cache-Clear](/assets/media/blog-images/2024-10-01-understanding-index-request-cache/keys_to_clean_delete_and_fetch.png){:class="img-centered" style="width:650px;"}

The following is the implementation of the `cleanCache` method:

```javascript
function cleanCache() {
    // Step 1: Initialize sets for keys to clean
    currentKeysToClean = new Set()
    currentFullClean = new Set()

    // Step 2: Process the list of keys that need to be cleaned
    for each cleanupKey in keysToClean {
        keysToClean.remove(cleanupKey)
        if (shard is closed or cacheClearAPI called) {
            currentFullClean.add(cleanupKey.entity.getCacheIdentity())
        } else {
            currentKeysToClean.add(cleanupKey)
        }
    }

    // Step 3: Process the cache and remove identified keys
    for each key in cache.keys() {
        if (currentFullClean.contains(key.entity.getCacheIdentity()) or 
            currentKeysToClean.contains(new CleanupKey(key.entity, key.readerCacheKey))) {
            cache.remove(key)
        }
    }

    // Step 4: Refresh the cache
    cache.refresh()
}
```

## Wrapping up

> Index request cache plays a crucial role in OpenSearch's efficiency. By understanding how it works, you can optimize performance and fine-tune your configurations with greater confidence.

Reviewer: ```suggestion
The index request cache plays a crucial role in ensuring the efficiency of OpenSearch. By understanding how it works, you can optimize performance and fine-tune your configurations with greater confidence.
```

> OpenSearch thrives on community contributions. If you have ideas or see opportunities for improvement, consider contributing. Your input can help shape the future of this search technology.

Reviewer: ```suggestion
The OpenSearch Project thrives on community contributions. If you have suggestions for improvement, consider contributing to the project. Your input can help shape the future of this search technology.
```

---
layout: post
title: "Optimizing query performance: The implementation of the OpenSearch index request cache" 
authors:
   - awskiran
   - kkhatua
   - upasagar
   - kolchfa

> date: 2024-10-01 

Reviewer 3: Update the publish date to today.

categories:
  - technical-posts
  - search
meta_keywords: OpenSearch cluster, Opensearch caching, index request cache, search performance optimization, search latency

> meta_description: 

Reviewer 3: Update meta keywords and description with the following: 

_meta_keywords:_   index request cache, search query performance, cache implementation, cache invalidation, cache cleanup, OpenSearch indices request cache

_meta_description:_ Explore the implementation of the OpenSearch index request cache and how this crucial mechanism enhances search query performance and efficiency.

## Understanding cache entries

Each cache entry is a key-value pair of **Key → BytesReference**.

A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:

> 1. **CacheEntity:** An [IndexShardCacheEntity](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesService.java#L1866C24-L1866C45) that contains the [IndexShard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java). This reference links the key to the shard it belongs to.

Author: An IndexShardCacheEntity that contains the IndexShard seems incorrect, should it be - "An IndexShardCacheEntity contains Index Shard" ?
